src/gentropy/dataset/colocalisation.py

```diff
@@ -23,3 +35,105 @@ def get_schema(cls: type[Colocalisation]) -> StructType:
+        if filter_by_qtl and filter_by_qtl not in valid_qtls:
+            raise ValueError(f"There are no studies with QTL type {filter_by_qtl}")
+
+        if filter_by_colocalisation_method not in [
+            "ECaviar",
+            "Coloc",
+        ]:  # TODO: Write helper class to retrieve coloc method names
+            raise ValueError(
+                f"Colocalisation method {filter_by_colocalisation_method} is not supported."
+            )
+
+        method_colocalisation_metric = ColocalisationStep._get_colocalisation_class(
+            filter_by_colocalisation_method
+        ).METHOD_METRIC  # type: ignore
+
+        coloc_filtering_expr = [
+            f.col("rightGeneId").isNull(),
+            f.col("colocalisationMethod") == filter_by_colocalisation_method,
+        ]
+        if filter_by_qtl:
+            coloc_filtering_expr.append(f.col("rightStudyType") == filter_by_qtl)
+
+        filtered_colocalisation = (
+            # Bring rightStudyType and rightGeneId and filter by rows where the gene is null,
+            # which is equivalent to filtering studyloci from gwas on the right side
+            self.append_right_study_metadata(
+                study_loci, study_index, ["studyType", "geneId"]
+            )
+            # it also filters based on method and qtl type
+            .filter(reduce(lambda a, b: a & b, coloc_filtering_expr))
+            # and filters colocalisation results to only include the subset of studylocus that contains gwas studylocusid
+            .join(
+                study_loci.df.selectExpr("studyLocusId as leftStudyLocusId"),
+                "leftStudyLocusId",
+            )
+        )
+
+        return get_record_with_maximum_value(
+            filtered_colocalisation.withColumnRenamed(
+                "leftStudyLocusId", "studyLocusId"
+            ).withColumnRenamed("rightGeneId", "geneId"),
+            ["studyLocusId", "geneId"],
+            method_colocalisation_metric,
+        )
+
+    def append_right_study_metadata(
+        self: Colocalisation,
+        study_loci: StudyLocus | L2GGoldStandard,
+        study_index: StudyIndex,
+        metadata_cols: list[str],
+    ) -> DataFrame:
+        """Appends metadata from the study in the right side of the colocalisation dataset.
+
+        Args:
+            study_loci (StudyLocus | L2GGoldStandard): Dataset containing study loci that links the colocalisation dataset and the study index via the studyId
+            study_index (StudyIndex): Dataset containing study index that contains the metadata
+            metadata_cols (list[str]): List of study columns to append
+
+        Returns:
+            DataFrame: Colocalisation dataset with appended metadata of the right study
+        """
+        # TODO: make this flexible to bring metadata from the left study (2 joins)
+        return self.df.join(
+            study_loci.df.selectExpr(
```

Comment:
Like this? I'm fine with that, probably it is more readable because it doesn't start operating with the left/right nomenclature until later

```
        return (
            # Annotate study loci with study metadata
            study_loci.df.select("studyLocusId", "studyId")
            .join(
                f.broadcast(study_index.df.select("studyId", *metadata_cols)), "studyId"
            )
            # Append that to the right side of the colocalisation dataset
            .selectExpr(
                "studyLocusId as rightStudyLocusId",
                *[f"{col} as right{col[0].upper() + col[1:]}" for col in metadata_cols],
            )
            .join(self.df, "rightStudyLocusId", "right")
        )
```

Reply:
Yes, but my key point was to add the `drop_duplicates(["studyLocusId"])`